cypress/support/api_commands.js

```diff
@@ -177,6 +177,44 @@ Cypress.Commands.add('apiAddUserToTeam', (teamId, userId) => {
     });
 });
 
+/**
+ * Join teammates directly via API
+ * This API assume that the user is logged in and has cookie to access
+ * @param {String} name - The name arg previously used during team-create (unique but not GUID)
+ * @param {Integer} size - The amount of users to join
+ * All parameter required
+ */
+Cypress.Commands.add('apiBatchJoinTeam', (name, size = 60) => {
+    return cy.getCookie('MMUSERID').then((cookie) => {
+        const admin = cookie.value;
+        cy.request({method: 'GET', url: `/api/v4/teams/name/${name}`}).then((teamResp) => {
+            const key = teamResp.body.id;
+            cy.request({method: 'GET', url: `/api/v4/users?per_page=${size}`}).then((userResp) => {
+                // Shape the users list into batch request body, skip the team admin.
+                const ulist = userResp.body.filter((u) => u.id !== admin).map((user) => {
+                    return {team_id: key, user_id: user.id, roles: 'system_user', delete_at: 0, scheme_user: true, scheme_admin: true, explicit_roles: 'system_user'};
+                });
+
+                // The API errors when the list is more than 20 users, so do chunks.
+                for (let i = ulist.length; i > 20; i -= 20) {
+                    const twenty = [];
+
+                    // pop 20 off from list
+                    for (let j = 0; j < 20; j++) {
+                        twenty.push(ulist.pop());
+                    }
+
+                    // Send request on twenty
+                    cy.request({method: 'POST', headers: {'X-Requested-With': 'XMLHttpRequest'}, url: `/api/v4/teams/${key}/members/batch`, body: twenty, qs: {team_id: key}});
+                }
+
+                // Make the remaining batch request.
+                return cy.request({method: 'POST', headers: {'X-Requested-With': 'XMLHttpRequest'}, url: `/api/v4/teams/${key}/members/batch`, body: ulist, qs: {team_id: key}});
```

Comment:
I'm seeing 3 endpoints here that we will probably use in the future.  Can we pull out those as 3 separate API commands and have those use for `cy.apiBatchJoinTeam`?  Like,
- `cy.apiAddUsersToTeam` --> `/api/v4/teams/${teamId}/members/batch`
- `cy.apiGetTeamByName` --> `/api/v4/teams/name/${name}`
- `cy.apiGetUsers` --> `/api/v4/users?per_page=${size}`

Reply:
Yes, I followed the post-message-as reference and added post_add_teammates / get_non_teammates. With the refactor, I was able to use existing .api* calls.

Reply:
Hmm, sorry for not being clear. I suggests not to do it as Cypress task but rather as (api) command.  So, in `api_commands.js`, we can add something like:
```javascript
// Todo: add comments
Cypress.Commands.add('apiGetUsersNotInTeam', (teamId, page = 0, perPage = 60) => {
    return cy.request({
        method: 'GET',
        url: `/api/v4/users?not_in_team=${teamId}&page=${page}&per_page=${perPage}`,
        headers: {'X-Requested-With': 'XMLHttpRequest'},
    });
});

// Todo: add comments
Cypress.Commands.add('apiAddUsersToTeam', (teamId, teamMembers) => {
    return cy.request({
        method: 'POST',
        url: `/api/v4/teams/${teamId}/members/batch`,
        headers: {'X-Requested-With': 'XMLHttpRequest'},
        body: teamMembers,
    });
});
```
which can be used like:
```javascript
cy.loginAsNewUser().then(() => {
// # Create new team and visit its URL
    cy.apiCreateTeam('test-team', 'Test Team').then((createResponse) => {
        const testTeam = createResponse.body;
        cy.visit(`/${testTeam.name}`);

        cy.apiGetUsersNotInTeam(testTeam.id).then((getResponse) => {
            const userIds = getResponse.body.map((user) => user.id);

            Cypress._.chunk(userIds, 20).forEach((batchedUserIds) => {
                const teamMembers = batchedUserIds.map((userId) => ({team_id: testTeam.id, user_id: userId}));
                cy.apiAddUsersToTeam(testTeam.id, teamMembers);
            });
        });
    });
});
```

Reply:
+1 to this.

Reply:
ok yes, I've removed the cy.task related logic. (I went down that path actually after trying custom commands, but then remembered the discussion around cy.task for the batch API in the UI Automation channel.)